### PR TITLE
Clustering

### LINKED ISSUE #10: Re-do clustering
From https://github.com/UCL/HHprY-Project/issues/18 :

> ACTION 4. Clustering: 
> 4A) overlapping hits should be clustered together no matter what the pSS, leaving non-overlapping ones separate. [...]
> 4B) Could the number of clusters in the default view be varied by length. We could have a rule of up to 10 clusters for all proteins less than 1000 residues, then 1 extra cluster per 200 residues on top. So for YLL040W/Vps13 (3144 aa) the page would have up to 21 clusters. Does not matter about haing 21 different colours, but we could try anyway – I could supply some more pastel colours. [...]

Actions:
- [x] @ilectra Make a new clustering algorithm (starting point, our methodology will most likely evolve over time until we get it right):

1) Remove hit that contains the same name as protein in its description, if such hit exists.
2) Start from hit with top probability and go down the list.
3) For every hit, compare with each hit with higher probability, and (percentages quoted are of the shorter of the two hits being compared)
  A) If <(max(10%,10)) residues overlapping, the hits are considered separate.
[Did I write this down correctly? For example, a hit of length 30, won't be considered overlapping with another if they overlap by 9 residues, even though this is almost 1/3 of the length of the short one?]
  B) If >50% overlap, the hits are considered overlapping, and they'll be added to the same cluster.
[What happens if there's legitimate overlap with more than one hit/cluster? - this can still be an issue for edge cases, despite point (4) below.]
4) If the hit overlapped with other(s) with percentage between 10% and 50%, cluster it together with the one with the highest overlap.
5) If the hit didn't overlap with any other, it starts it own cluster.
6) At the end, each "cluster" is a new, longer "hit", containing all the overlapping ones.
[Order of transversal matters here - are clusters "creeping" to one side correct? Example:]
```
Hits:
---------
     --------
         ------------
Are clustered as:
---------------------
even though there's no overlap between the top and bottom one.
```
7) As a result of the previous procedure, anything contained fully inside something bigger, will be clustered together.
8) Keep name and probability of top hit, number of hits clustered together
9) If hits with the same name appear, cluster them together.
[Do we really need to/can do this?]
10) The procedure should be able to handle both pfam (minimal overlaps) and pdb (tons of overlapping hits).

- [ ] @timlevine provide list of proteins with different features, to test the clustering algorithm.

### PR DESCRIPTION
**Pull request by ilectra**
Fixes #10

### CODE REVIEW

File: clustering.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+
+    def _clean(self):
+        self._x1cl_min = []
+        self._x1cl_max = []
+        self._x2cl_min = []
+        self._x2cl_max = []
+        self._ncl = []
+        self._clabels = []
+
+    # Given lists of hit beginings and ends, fill up the cluster lists data members,
```
[the remark this answers is not in the document]
Comment (ilectra) [reply clustering.py]: Can I leave converting the comments to docstrings for later, whenever I fix issue #23 ?

[the remark this answers is not in the document]
Comment (jonc125) [reply clustering.py]: Sure :)

File: clustering.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+    ZERO = auto()
+
+
+class Clustering:
+    def __init__(self, ov_min=10, ov_min1=0.2, ov_min2=0.5):
+        self._x1cl_min = []
+        self._x1cl_max = []
+        self._x2cl_min = []
+        self._x2cl_max = []
+        self._ncl = []
```
[the remark this answers is not in the document]
Comment (ilectra) [reply clustering.py]: Good point, I changed all of them.

File: clustering.py [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+                self._ncl[i] += 1
+                return i
+        self._x1cl_min.append(x1)
+        self._x1cl_max.append(x1)
+        self._x2cl_min.append(x2)
+        self._x2cl_max.append(x2)
+        self._ncl.append(1)
+        return len(self._ncl)-1
+
+    # Check if hit with limits [x1,x2] overlaps with index-th cluster.
```
[the remark this answers is not in the document]
Comment (ilectra) [reply clustering.py]: Done - let me know if it makes sense!

**Comment by jonc125** [comment]
Yep, looks good to me!

**Pull request merged**